HIP/hip-1056.md

```diff
@@ -0,0 +1,1952 @@
+   threshold of stake weights, so that I can cost-effectively confirm that the stream represents the consensus output of
+   the network.
+4. As a downstream developer, I want a low latency feed that provides the state changes associated with transaction(s)
+   within a block, so that I can interact with state changes with the smallest possible delay.
+
+## Specification
+
+The Block Stream will be defined by the continuous transmission of Block information defined as follows.
+- Each `Block` represents all inputs and outputs of the consensus network for zero or more whole `Rounds`. The maximum
+  number of rounds per block will be configurable and start at 1 round per block. It could be increased later if
+  necessary for performance reasons or if rounds become very frequent and small.
+    - A block *may* be zero rounds because the last block prior to a network "freeze" must communicate the state hash
+      (via a block proof) of the state immediately prior to network shutdown. This is only possible with an extra empty
+      block because the state hash in each block proof is the state hash at the *beginning* of that block/round. This
+      delay of state hash communication is required to simultaneously meet Hedera performance goals and also meet EVM
+      expectations for immediate access to recent (the last 256) block hash values.
+- A `Round` contains a sequence of zero or more `Events` that have reached consensus at the same time. Round length in real
+  world time depends on internet latency and event creation rate. In Hedera mainnet today it is around 1 round per
+  second. In the future this could be shorter to lower latency, but would never make sense to be less than ping time.
+- An `Event` contains a sequence of zero or more `EventTransactions` submitted by a single node.
+- Each `EventTransaction` is either a `SignedTransaction` or a system transaction.
+- A `SignedTransaction` is transmitted as a byte array, contains the signed transaction bytes of a user-submitted
+  transaction exactly as processed by the network, and is followed by transaction outputs and state changes resulting
+  from that transaction.
+- System transactions are represented by explicitly defined protobuf messages and used for consensus node to consensus
+  node communication.
+
+These concepts are core to the Hashgraph consensus mechanism and have direct impacts on the structure of data in the 
+streams.
+
+Block streams are built out of Items. Those items can be delivered in a continuous stream containing one or more blocks.
+Or they can be delivered as a single block using the `Block` message type container. If a single block is delivered as
+a file it must be either a raw protobuf `Block` message with file name of `<BLOCK_NUMBER>.blk` e.g.
+`0000000000000000001.blk`. The file name contains a total of 19 digits to allow for 2^63 blocks. The file format may
+also be a compressed file e.g. `0000000000000000001.blk.gz` or `0000000000000000001.blk.ztsd`.
```

Comment:
This is ambiguous. The HIP presents three possible names for the block file in the bucket: 0000000000000000001.blk, 0000000000000000001.blk.gz, or 0000000000000000001.blk.zst. Since the filename is the API for how mirror nodes will consume it from the bucket, we have to explicitly pick one of these. Consensus nodes will only upload one of these three formats and if mirror nodes pick the wrong format then it won't work or it has to waste effort and try all 3 names.

Reply:
The file name _is not_ an API.  We'll work out the _temporary_ storage of files in buckets independent of the Block Stream HIP, as the HIP _should_ neither mandate nor envision file-in-bucket storage as a long-term solution.

Reply:
I just got out of a meeting where we discussed phase 1 being block streams in the bucket as the canonical source. So the file name will very much be an API if that's the case.

Reply:
I would say we need to reconsider that, as the block stream is _not a file_, definitionally.
If we want a _separate_ HIP, that offers _temporary_ public access, then that might make sense, but making a temporary transitional access part of the permanent public API is not what I would consider well designed.

Reply:
I believe there's a needed separation of the API the HIP is proposing and the API the team requires to achieve this goal in the interim to meet a schedule.
MNs consuming buckets is a project phasing optimization to get this out soon but is not an intention of this HIPs specification nor of long term usage. 
For instance a non Hedera run Mirror Node would consider this HIP functional and implemented when they can read from the block stream and should hopefully never have to pull from the cloud bucket.
However, for the project I do believe we must commit to one file format so MNs early on can commit to one format.

Reply:
Alternatively, @steven-sheehy's suggestion is fair. 
Assuming the suggestion is CN will really have 2 output formats for Block Items ( (i) block streams and ii) BlockFile for special cases) and we should specify both for clarity.
@jasperpotts what do you think?

Reply:
Will be a separate HIP.